lib/PlatoonService/src/V2VCommManager.cpp

```diff
@@ -0,0 +1,534 @@
+            else
+            {
+                LOG_DEBUG("All participants responded to heartbeat.");
+                m_v2vStatus = V2V_STATUS_OK;
+            }
+
+            /* Stop timer. */
+            m_vehicleHeartbeatTimeoutTimer.stop();
+        }
+    }
+
+    return m_v2vStatus;
+}
+
+bool V2VCommManager::sendWaypoint(const Waypoint& waypoint)
+{
+    bool   isSuccessful = false;
+    String payload;
+    waypoint.serialize(payload);
+
+    if (true == payload.isEmpty())
+    {
+        LOG_DEBUG("Failed to serialize waypoint.");
+    }
+    else if (false == m_mqttClient.publish(m_waypointOutputTopic, false, payload))
+    {
+        LOG_ERROR("Failed to publish MQTT message to %s.", m_waypointOutputTopic.c_str());
+    }
+    else
+    {
+        isSuccessful = true;
+    }
+
+    return isSuccessful;
+}
+
+bool V2VCommManager::getNextWaypoint(Waypoint& waypoint)
+{
+    bool isSuccessful = false;
+
+    if (false == m_waypointQueue.empty())
+    {
+        /* Retrieve next waypoint. */
+        Waypoint* nextWaypoint = m_waypointQueue.front();
+        m_waypointQueue.pop();
+
+        /* Copy waypoint. */
+        waypoint = *nextWaypoint;
+
+        /* Delete queued waypoint. */
+        delete nextWaypoint;
+
+        isSuccessful = true;
+    }
+
+    return isSuccessful;
+}
+
+size_t V2VCommManager::getWaypointQueueSize() const
+{
+    return m_waypointQueue.size();
+}
+
+/******************************************************************************
+ * Protected Methods
+ *****************************************************************************/
+
+/******************************************************************************
+ * Private Methods
+ *****************************************************************************/
+
+void V2VCommManager::targetWaypointTopicCallback(const String& payload)
+{
+    Waypoint* waypoint = Waypoint::deserialize(payload);
+
+    if (nullptr == waypoint)
+    {
+        LOG_ERROR("Failed to deserialize received waypoint.");
+    }
+    else
+    {
+        m_waypointQueue.push(waypoint);
+    }
+}
+
+void V2VCommManager::platoonHeartbeatTopicCallback(const String& payload)
+{
+    /* Deserialize payload. */
+    StaticJsonDocument<JSON_HEARTBEAT_MAX_SIZE> jsonPayload;
+    DeserializationError                        error = deserializeJson(jsonPayload, payload.c_str());
+
+    if (DeserializationError::Ok != error)
+    {
+        LOG_ERROR("JSON Deserialization Error %d.", error);
+    }
+    else
+    {
+        /* Send vehicle heartbeat. */
+        JsonVariant                                 jsonTimestamp = jsonPayload["timestamp"]; /* Timestamp [ms]. */
+        StaticJsonDocument<JSON_HEARTBEAT_MAX_SIZE> heartbeatDoc;
+        String                                      heartbeatPayload;
+
+        if (false == jsonTimestamp.isNull())
+        {
+            heartbeatDoc["id"]        = m_vehicleId;
+            heartbeatDoc["timestamp"] = jsonTimestamp.as<uint32_t>();

```

Comment:
Please add a comment why its important to send the timestamp back.

Reply:
Done.